NitroxClient/GameLogic/Bases/GhostMetadataApplier.cs

```diff
@@ -0,0 +1,209 @@
+using NitroxClient.Unity.Helper;
+using NitroxModel.DataStructures.GameLogic.Entities.Metadata;
+using NitroxModel.DataStructures.GameLogic.Entities.Metadata.Bases;
+using NitroxModel_Subnautica.DataStructures;
+using System.Collections;
+using UnityEngine;
+using UWE;
+
+namespace NitroxClient.GameLogic.Bases;
+
+public static class GhostMetadataApplier
+{
+    public static T From<T>(BaseGhost baseGhost) where T : GhostMetadata, new()
+    {
+        T metadata = new()
+        {
+            TargetOffset = baseGhost.targetOffset.ToDto()
+        };
+        return metadata;
+    }
+
+    public static void ApplyBasicMetadataTo(this GhostMetadata ghostMetadata, BaseGhost baseGhost)
+    {
+        baseGhost.targetOffset = ghostMetadata.TargetOffset.ToUnity();
+    }
+
+    public static GhostMetadata GetMetadataForGhost(BaseGhost baseGhost)
+    {
+        // Specific case in which a piece was deconstructed and resulted in a BaseDeconstructable with a normal BaseGhost
+        if (BuildUtils.IsUnderBaseDeconstructable(baseGhost))
+        {
+            return GetBaseDeconstructableMetadata(baseGhost);
+        }
+
+        GhostMetadata metadata = baseGhost switch
+        {
+            BaseAddWaterPark or BaseAddPartitionDoorGhost or BaseAddModuleGhost or BaseAddFaceGhost => GetBaseAnchoredFaceMetadata(baseGhost),
+            BaseAddPartitionGhost => GetBaseAnchoredCellMetadata(baseGhost),
+            _ => From<GhostMetadata>(baseGhost),
+        };
+        return metadata;
+    }
+
+    public static IEnumerator ApplyMetadataToGhost(BaseGhost baseGhost, EntityMetadata entityMetadata, Base @base)
+    {
+        if (entityMetadata is not GhostMetadata ghostMetadata)
+        {
+            Log.Error($"Trying to apply metadata to a ghost that is not of type {nameof(GhostMetadata)} : [{entityMetadata.GetType()}]");
+            yield break;
+        }
+
+        if (BuildUtils.IsUnderBaseDeconstructable(baseGhost, true) && entityMetadata is BaseDeconstructableGhostMetadata deconstructableMetadata)
+        {
+            yield return deconstructableMetadata.ApplyBaseDeconstructableMetadataTo(baseGhost, @base);
+            yield break;
+        }
+
+        switch (baseGhost)
+        {
+            case BaseAddWaterPark:
+            case BaseAddPartitionDoorGhost:
+            case BaseAddModuleGhost:
+            case BaseAddFaceGhost:
+                if (ghostMetadata is BaseAnchoredFaceGhostMetadata faceMetadata)
```

Comment:
Can it happen that baseGhost and Metadata types don't align? If so I would like that we log this.